File: functions/shared/src/github.ts
/** GitHub API and webhook models (Zod schemas). */

import { z } from "zod";

/** Repository reference from webhook payload or GitHub API. */
export const RepoRefSchema = z.object({
  full_name: z.string(),
  id: z.number(),
});
export type RepoRef = z.infer<typeof RepoRefSchema>;

/** Installation reference from webhook payload. */
export const GitHubInstallationRefSchema = z.object({
  id: z.number(),
});
export type GitHubInstallationRef = z.infer<typeof GitHubInstallationRefSchema>;

/** GitHub App installation_repositories webhook payload. */
export const GitHubWebhookPayloadSchema = z.object({
  action: z.string(),
  installation: GitHubInstallationRefSchema,
  repositories_added: z.array(RepoRefSchema).default([]),
  repositories_removed: z.array(RepoRefSchema).default([]),
});
export type GitHubWebhookPayload = z.infer<typeof GitHubWebhookPayloadSchema>;

/** GitHub OAuth token exchange response. */
export const GitHubTokenResponseSchema = z.object({
  access_token: z.string(),
  token_type: z.string().default("bearer"),
  scope: z.string().optional(),
  expires_in: z.number().optional(),
});
export type GitHubTokenResponse = z.infer<typeof GitHubTokenResponseSchema>;

/** GitHub user from /user API. */
export const GitHubUserSchema = z.object({
  id: z.number(),
  login: z.string(),
  email: z.string().optional(),
});
export type GitHubUser = z.infer<typeof GitHubUserSchema>;

/** Repository access from GitHub App installation. */
export const RepoAccessSchema = z.object({
  full_name: z.string(),
  id: z.number(),
});
export type RepoAccess = z.infer<typeof RepoAccessSchema>;

/** GitHub App installation (minimal for filtering). */
export const GitHubInstallationSchema = z.object({
  id: z.number(),
  app_id: z.number().optional(),
});
export type GitHubInstallation = z.infer<typeof GitHubInstallationSchema>;

/** GitHub /user/emails item. */
export const GitHubEmailSchema = z.object({
  email: z.string(),
  primary: z.boolean().optional(),
  verified: z.boolean().optional(),
});
export type GitHubEmail = z.infer<typeof GitHubEmailSchema>;

/** GitHub /user/installations response. */
export const GitHubInstallationsResponseSchema = z.object({
  installations: z.array(GitHubInstallationSchema).default([]),
});
export type GitHubInstallationsResponse = z.infer<
  typeof GitHubInstallationsResponseSchema
>;

/** GitHub /user/installations/{id}/repositories response. */
export const GitHubRepositoriesResponseSchema = z.object({
  repositories: z.array(RepoAccessSchema).default([]),
});
export type GitHubRepositoriesResponse = z.infer<
  typeof GitHubRepositoriesResponseSchema
>;

File: functions/auth_callback/shared/src/functions_shared/github.py
"""GitHub API response models for auth functions."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class GitHubTokenResponse(BaseModel):
    """GitHub OAuth token exchange response."""

    access_token: str
    token_type: str = "bearer"
    scope: str | None = None
    expires_in: int | None = None

    model_config = ConfigDict(extra="ignore")


class GitHubUser(BaseModel):
    """GitHub user from /user API."""

    id: int
    login: str
    email: str | None = None

    model_config = ConfigDict(extra="ignore")


class RepoAccess(BaseModel):
    """Repository access from GitHub App installation."""

    full_name: str = Field(description="e.g. owner/repo")
    id: int = Field(description="GitHub repo ID")

    model_config = ConfigDict(extra="ignore")


class GitHubInstallation(BaseModel):
    """GitHub App installation (minimal for filtering)."""

    id: int
    app_id: int | None = None

    model_config = ConfigDict(extra="ignore")
